src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly legacy!: LegacyServiceSetup;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: Record<string, any>;
+  private readonly plugins!: PluginsServiceSetup;
+  private readonly template = compileFile(join(__dirname, 'views/ui_app.pug'));
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
+
+  private getCapabilities() {
+    const modifiers = this.http.server.getCapabilitiesModifiers();
+    const defaultCapabilties = this.http.server.getDefaultCapabilities();
+    const capabilities = mergeCapabilities(defaultCapabilties, {
+      // Get legacy nav links
+      navLinks: Object.assign(
+        {},
+        ...this.http.server.getUiNavLinks().map(({ _id }) => ({ [_id]: true }))
+      ),
+    });
+
+    return modifiers.reduce(
+      async (resolvedCapabilties, modifier) => modifier(this.request, await resolvedCapabilties),
+      Promise.resolve(capabilities)
+    );
+  }
+
+  private async getVariables(app: App, basePath: string, injectedOverrides: Record<string, any>) {
+    const providers: Provider[] = (this.legacy.uiExports as any).defaultInjectedVarProviders || [];
+    const replacers: Replacer[] = (this.legacy.uiExports as any).injectedVarsReplacers || [];
+    const defaultInjectedVars = providers.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(
+          defaults,
+          fn(this.http.server, pluginSpec.readConfigValue(this.http.server.config, []))
+        ),
+      {}
+    );
+    const appInjectedVars = await this.http.server.getInjectedUiAppVars(app.getId());
+    const injectedVars = mergeVariables(defaultInjectedVars, appInjectedVars, injectedOverrides);
+    const vars = await replacers.reduce(
+      async (variables, replacer) => await replacer(variables, this.request, this.http.server),
+      Promise.resolve(injectedVars)
+    );
+
+    return {
+      strictCsp: this.http.csp.strict,
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${app.getId()}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: this.uiSettings?.user?.['theme:darkMode']?.userValue ?? false,
+      injectedMetadata: {
+        version: this.http.server.version,
+        buildNumber: this.env.packageInfo.buildNum,
+        branch: this.env.packageInfo.branch,
+        basePath,
+        env: this.env,
+        legacyMode: app.getId() !== 'core',
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: {
+          warnLegacyBrowsers: this.http.csp.warnLegacyBrowsers,
+        },
+        vars,
+        uiPlugins: [...this.plugins.uiPlugins.public].map(([id, plugin]) => ({ id, plugin })),
+        legacyMetadata: {
```

Comment:
Question: do we need legacy metadata in the NP? Probably only some of them? `uiSettings`, for example.

Reply:
We have to provide this since injected metadata has an API fronting this. Once legacy is removed I think we can remove this.